Sources/Utils.php

```diff
@@ -1053,48 +1053,96 @@ public static function buildRegex(array $strings, ?string $delim = null, bool $r
 
 		// Can we trim common characters from the end?
 		$trailing = '';
+		$i = -1;
 		unset($normalized_strings);
 
-		while (mb_strlen($strings[0], $encoding) > 1) {
-			$last_char = mb_substr($strings[0], -1, 1, $encoding);
+		while (\strlen($strings[0]) > 1) {
+			$last_char = mb_substr($strings[0], $i, null, $encoding);
 
 			foreach ($strings as $string) {
 				if (!str_ends_with($string, $last_char)) {
 					break 2;
 				}
 			}
 
-			$strings = array_map(fn($string) => mb_substr($string, 0, -1, $encoding), $strings);
-			$trailing = $last_char . $trailing;
+			$i--;
+			$trailing = $last_char;
 		}
 
-		// Create the trie from the strings.
+		if ($trailing !== '') {
+			$strings = array_map(fn($string) => substr($string, 0, -\strlen($trailing)), $strings);
+		}
+
+		// Build a radix tree from the input strings
 		$trie = [];
 
 		foreach ($strings as $string) {
-			$chars = mb_str_split($string, 1, $encoding);
+			$current_node = &$trie;
+
+			$remaining = $string;
+
+			while ($remaining !== '') {
+				$matched = false;
+
+				// Check each existing branch at this level
+				foreach ($current_node as $key => &$subtree) {
+					// Find longest common prefix
+					$len = 0;
+					$max = min(mb_strlen($key, $encoding), mb_strlen($remaining, $encoding));
+					while (
+						$len < $max &&
+						mb_substr($key, $len, 1, $encoding) === mb_substr($remaining, $len, 1, $encoding)
+					) {
+						$len++;
+					}
+
+					if ($len === 0) {
+						continue; // no prefix match, try next branch
+					}
+
+					$prefix = mb_substr($key, 0, $len, $encoding);
+					$key_remainder = mb_substr($key, $len, null, $encoding);
+					$remaining_remainder = mb_substr($remaining, $len, null, $encoding);
```

Comment:
I have some code which cuts these string based on byte boundaries,   but that will make this function UTF-8 only. Would that be acceptable?